### PR TITLE
improved SkinGUI for fingertip 

### PR DESCRIPTION
**Pull request by simeonedussoni**
implemented:

- skin type `fingertip 4L/R` for mk1.1 PCB testing 
- skin type `fingertipMID` for 3DMID visualization 
- palm workaround (added a `fakePalm` with just one virtual taxel with zero gain and size to avoid crashing of the GUI)

### CODE REVIEW

**Comment by simeonedussoni** [comment]
Need to update also the CMakeList file.

File: src/tools/iCubSkinGui/plugin/CMakeLists.txt [hunk 1]
```diff
@@ -30,7 +34,10 @@ set(QtICubSkinGuiPlugin_HDRS qticubskinguiplugin.h
                              include/Fingertip2Right.h
                              include/Fingertip3Left.h
                              include/Fingertip3Right.h
+                             include/Fingertip4Left.h
+                             include/Fingertip4Right.h
                              include/Fingertip.h
+                             include/fakePalm.h
```
Comment (pattacini) [inline src/tools/iCubSkinGui/plugin/CMakeLists.txt:40]: The inclusion of `FingertipMID.h` is missing.

File: src/tools/iCubSkinGui/plugin/include/SkinMeshThreadCan.h [hunk 2]
```diff
@@ -141,6 +163,14 @@ class SkinMeshThreadCan : public PeriodicThread
...
                         {
                             sensor[id]=new Fingertip3R(xc,yc,th,gain,layoutNum,lrMirror);
+                        }
+                         if (type=="fingertip4R")
+                        {
+                            sensor[id]=new Fingertip4R(xc,yc,th,gain,layoutNum,lrMirror);
+                        }
+                         if (type=="fingertipMID")
+                        {
+                            sensor[id]=new FingertipMID(xc,yc,th,gain,layoutNum,lrMirror);
```
Comment (pattacini) [inline src/tools/iCubSkinGui/plugin/include/SkinMeshThreadCan.h:173]: Better off turning this check that is getting too long into a chain of `else-if` as already done elsewhere within this codebase.

**Review by pattacini — CHANGES_REQUESTED** [review]
Hi @simeonedussoni 

A couple of inline comments to address.

A few more comments:
- Do we need also configuration files?
- If you did some tests like showing the results on the GUI, would you be willing to post them here?

**Comment by simeonedussoni** [comment]
> Hi @simeonedussoni
> 
> A couple of inline comments to address.
> 
> A few more comments:
> 
>     * Do we need also configuration files?
> 
>     * If you did some tests like showing the results on the GUI, would you be willing to post them here?

here we have an example of configuration file showing all 5 types of layout, one of which activated by touchin the PCB

immagine

concerning the configuration files:
- for the (soon) mounted icub hands they are already in the repo under `/app/skinGui/conf/skinGui/left_hand_V2_2.ini`
- for ergocub: they are still to be produced, I can put there in the same location with "speaking names" and add to the PR
- for using the unwrapped version of the layout will be deployed in the `test-skin-patches` repo to keep them separate from the actual robot patches

**Comment by pattacini** [comment]
>for ergocub: they are still to be produced, I can put there in the same location with "speaking names" and add to the PR

This repo contains icub-specific SW. A better place is definitely https://github.com/icub-tech-iit/ergocub-software. I would wait for @Nicogene to show up before proceeding in this respect though.

**Comment by pattacini** [comment]
>for using the unwrapped version of the layout will be deployed in the test-skin-patches repo to keep them separate from the actual robot patches

Please, get aligned with @davidetome for that.

File: src/tools/iCubSkinGui/plugin/include/SkinMeshThreadCan.h (hunk 2 above)
Comment (simeonedussoni) [reply src/tools/iCubSkinGui/plugin/include/SkinMeshThreadCan.h]: 2202a362d7f76fd55126b9505dcd116b541d1d50 resolved both comments here and above

Comment (pattacini) [reply src/tools/iCubSkinGui/plugin/include/SkinMeshThreadCan.h]: Thanks!

**Review by pattacini — APPROVED** [review]
Awaiting the CI before merging.

**Comment by maggia80** [comment]
@simeonedussoni a few comments:

please check the author and the copyright of the committed files. 

add somewhere an explanation about which `gui ` corresponds to which fingertip and in which configuration (wrapped or unwrapped).

**Comment by Nicogene** [comment]
> > for ergocub: they are still to be produced, I can put there in the same location with "speaking names" and add to the PR
> 
> This repo contains icub-specific SW. A better place for the ergoCub conf files is definitely [icub-tech-iit/ergocub-software](https://github.com/icub-tech-iit/ergocub-software). I would wait for @Nicogene to show up before proceeding in this respect though.

Yes but I have no idea how can the SkinGui load fingertip compiled externally, if they are loaded run-time I am ok to move in ergocub-software

**Comment by pattacini** [comment]
> Yes but I have no idea how can the SkinGui load fingertip compiled externally, if they are loaded run-time I am ok to move in ergocub-software

I meant only the .ini files.

**Pull request merged**